temporal-sdk/src/main/java/io/temporal/workflow/UpdateMethod.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (C) 2023 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.workflow;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Indicates that the method is a update handler method. A update method gets executed when a
+ * workflow receives a update after the validator is called.
+ *
+ * <p>This annotation applies only to workflow interface methods.
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface UpdateMethod {
```

Comment:
I wonder if it's clearer to just have an update validator annotation than a validator class. This annotation can _require_ a name, and we can validate at registration time (along w/ our other annotation validations) that the name references an update and that the arguments are exactly the same and return type is void. This is better than `Object[]` arguments.

I think this:

```java
public class MyWorkflow {
  private String thing = "<unset>;

  // ...

  @UpdateMethod
  public String doUpdate(int someArg) {
    // ...
  }

  @UpdateValidationMethod("doUpdate")
  public void validateUpdate(int someArg) {
    Preconditions.checkArgument(someArg >= 0, "Argument can't be negative");
    Preconditions.checkState(!"<unset>".equals(thing), "Thing not ready");
  }
}
```

Looks better than:

```java
public class MyWorkflow {
  private String thing = "<unset>;

  // ...

  @UpdateMethod(validator=DoUpdateValidator.class)
  public String doUpdate(int someArg) {
    // ...
  }

  public static class DoUpdateValidator implements Validator<MyWorkflow> {
    public void apply(MyWorkflow workflowObject, Object[] args) {
      Preconditions.checkArgument((int)args[0] >= 0, "Argument can't be negative");
      Preconditions.checkState(!"<unset>".equals(MyWorkflow.thing), "Thing not ready");
    }
  }
}
```

Reply:
That was my initial idea as well. @Spikhalskiy argued the validator approach was the more common Java pattern and we should support it, but we can add other in the future.

Reply:
I think we have to weigh common Java pattern with losing argument types and general user annoyance. I would rather not have to create a separate class and cast all my arguments manually even if that's the "common Java pattern".